File: src/MyServiceBus/Persistence/OutboxSendEndpoint.cs
using System.Diagnostics;
using MyServiceBus.Serialization;

namespace MyServiceBus.Persistence;

internal sealed class OutboxSendEndpoint : ISendEndpoint
{
    private readonly OutboxSession session;
    private readonly ISendEndpoint fallback;
    private readonly ISendPipe sendPipe;
    private readonly IMessageSerializer serializer;
    private readonly Uri destination;
    private readonly Uri source;
    private readonly ISendContextFactory contextFactory;
    private readonly Action? ensureStarted;

    public OutboxSendEndpoint(
        OutboxSession session,
        ISendEndpoint fallback,
        ISendPipe sendPipe,
        IMessageSerializer serializer,
        Uri destination,
        Uri source,
        ISendContextFactory contextFactory,
        Action? ensureStarted)
    {
        this.session = session;
        this.fallback = fallback;
        this.sendPipe = sendPipe;
        this.serializer = serializer;
        this.destination = destination;
        this.source = source;
        this.contextFactory = contextFactory;
        this.ensureStarted = ensureStarted;
    }

    public Task Send<T>(T message, Action<ISendContext>? contextCallback = null, CancellationToken cancellationToken = default)
        where T : class => Send<T>((object)message, contextCallback, cancellationToken);

    public async Task Send<T>(object message, Action<ISendContext>? contextCallback = null, CancellationToken cancellationToken = default)
        where T : class
    {
        if (session.Writer is not { } writer)
        {
            await fallback.Send<T>(message, contextCallback, cancellationToken);
            return;
        }

        ensureStarted?.Invoke();
        var context = contextFactory.Create(MessageTypeCache.GetMessageTypes(typeof(T)), serializer, cancellationToken);
        context.MessageId = Guid.NewGuid().ToString();
        context.SourceAddress = source;
        context.DestinationAddress = destination;
        contextCallback?.Invoke(context);

        await sendPipe.Send(context);
        var typed = message is T value ? value : (T)MessageProxy.Create(typeof(T), message);
        await writer.AddAsync(OutboxMessageFactory.Create(typed, context), cancellationToken);
    }
}

File: src/Java/myservicebus/src/main/java/com/myservicebus/persistence/OutboxSendEndpoint.java
package com.myservicebus.persistence;

import com.myservicebus.MessageUrn;
import com.myservicebus.SendContext;
import com.myservicebus.SendContextFactory;
import com.myservicebus.SendEndpoint;
import com.myservicebus.SendPipe;
import com.myservicebus.serialization.MessageSerializer;
import com.myservicebus.tasks.CancellationToken;
import java.net.URI;
import java.util.concurrent.CompletableFuture;

public final class OutboxSendEndpoint implements SendEndpoint {
    private final OutboxSession session;
    private final SendEndpoint fallback;
    private final SendPipe sendPipe;
    private final MessageSerializer serializer;
    private final URI destination;
    private final URI source;
    private final SendContextFactory contextFactory;
    private final Runnable ensureStarted;

    public OutboxSendEndpoint(
            OutboxSession session,
            SendEndpoint fallback,
            SendPipe sendPipe,
            MessageSerializer serializer,
            URI destination,
            URI source,
            SendContextFactory contextFactory,
            Runnable ensureStarted) {
        this.session = session;
        this.fallback = fallback;
        this.sendPipe = sendPipe;
        this.serializer = serializer;
        this.destination = destination;
        this.source = source;
        this.contextFactory = contextFactory;
        this.ensureStarted = ensureStarted;
    }

    @Override
    public <T> CompletableFuture<Void> send(T message, CancellationToken cancellationToken) {
        SendContext context = contextFactory.create(message, cancellationToken);
        return capture(context);
    }

    @Override
    public CompletableFuture<Void> send(SendContext context) {
        return capture(context);
    }

    private CompletableFuture<Void> capture(SendContext context) {
        try {
            OutboxWriter writer = session.getWriter();
            if (writer == null) {
                return fallback.send(context);
            }
            ensureStarted.run();
            context.setSourceAddress(source);
            context.setDestinationAddress(destination);
            context.setMessageTypes(MessageUrn.forMessageTypes(context.getMessage().getClass()));
            return sendPipe.send(context).thenCompose(ignored -> {
                try {
                    return writer.add(
                            OutboxMessageFactory.create(context, serializer),
                            context.getCancellationToken());
                } catch (Exception failure) {
                    return CompletableFuture.failedFuture(failure);
                }
            });
        } catch (Exception failure) {
            return CompletableFuture.failedFuture(failure);
        }
    }
}
